src/openvic-simulation/scripts/Condition.hpp

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include "openvic-simulation/types/IdentifierRegistry.hpp"
+#include <variant>
+#include <vector>
+
+namespace OpenVic {
+	struct ConditionManager;
+
+	// Condition: the definition of an individual condition or scope
+	// AND, OR and NOT are conditions with complex type
+	// Scopes are treated as ANDs
+	struct Condition : HasIdentifier {
+		friend struct ConditionManager;
+
+		enum class value_type_t {
+			IDENTIFIER,
+			STRING,
+			BOOLEAN,
+			INTEGER,
+			REAL,
+			COMPLEX
+		};
+
+		enum class scope_t {
+			GLOBAL,
+			COUNTRY,
+			PROVINCE,
+			POP
+		};
+
+	private:
+		const value_type_t PROPERTY(value_type);
+		std::set<scope_t> PROPERTY(scopes);
+
+		Condition(std::string_view new_identifier, value_type_t new_value_type, std::set<scope_t>&& new_scopes);
+
+	public:
+		Condition(Condition&&) = default;
+	};
+
+	// ConditionNode: represents a node, which is a map of individual conditions associated with a value, tied by
+	// a logical relationship. Simple conditions are considered to be single-element ANDs.
+	struct ConditionNode {
+		using string_t = std::string;
+		using boolean_t = bool;
+		using integer_t = int32_t;
+		using real_t = fixed_point_t;
+		//TODO add complex types
+		using condition_set_t = std::set<ConditionNode>;
+		using value_t = std::variant<string_t, boolean_t, integer_t, real_t, condition_set_t>;
+
+	private:
+		Condition const* PROPERTY(condition);
+		value_t PROPERTY(value);
+
+		ConditionNode(Condition const* new_condition, value_t&& new_value);
+	};
+
+	struct ConditionManager {
+
+	private:
+		IdentifierRegistry<Condition> IDENTIFIER_REGISTRY(condition);
+
+		bool add_condition(std::string_view identifier, Condition::value_type_t value_type, Condition::scope_t scope);
+		bool setup_conditions();
+
+	public:
+		template<typename T> //TODO replace this with an expect
+		constexpr auto get_type_expect_callback(Condition::value_type_t t) {
+			switch (t) {
+				case Condition::value_type_t::IDENTIFIER: return NodeTools::expect_identifier;
+				case Condition::value_type_t::STRING: return NodeTools::expect_string;
+				case Condition::value_type_t::BOOLEAN: return NodeTools::expect_bool;
+				case Condition::value_type_t::INTEGER: return NodeTools::expect_int;
+				case Condition::value_type_t::REAL: return NodeTools::expect_fixed_point;
+				case Condition::value_type_t::COMPLEX: return this->expect_condition_dictionary;
```

Comment:
This confused me, I expected it to cause a build error, but of course it doesn't because the template isn't used anywhere, so the code isn't actually compiled.